Production file-based catalog specific for OCP versions

mgencur opened the pull request:
https://issues.redhat.com/browse/SRVCOM-3369

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

<!-- Please categorize your changes:
- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic
-->

-
-
-

creydr commented on .github/workflows/validate.yaml:
```diff
@@ -30,6 +30,12 @@ jobs:
 
+      - name: Install OPM
+        run: |
+          wget https://github.com/operator-framework/operator-registry/releases/download/v1.47.0/linux-amd64-opm && \
+            chmod +x linux-amd64-opm && \
+            mv linux-amd64-opm /usr/local/bin/opm
```
alternatively you can also use `go install`, so you don't have to care about arch & path:
`go install github.com/operator-framework/operator-registry/cmd/opm@v1.47.0`

pierDipi commented on .github/workflows/validate.yaml:
```diff
@@ -60,7 +66,10 @@ jobs:
+        env:
+          REGISTRY_REDHAT_IO_USERNAME: ${{ secrets.REGISTRY_REDHAT_IO_USERNAME }}
+          REGISTRY_REDHAT_IO_PASSWORD: ${{ secrets.REGISTRY_REDHAT_IO_PASSWORD }}
+        run: make generated-files generate-catalog
```
I'd include `generate-catalog` in `generated-files` so that others don't need to learn about this new command to run ?

mgencur replied:
Pushed the split.

pierDipi replied:
wow 1/2 hour

mgencur replied:
Alright. In github action it's much faster. About 5 minutes:
```
11:28:54.752 INFO:    Generating catalog for OCP 4.13
time="2024-10-11T11:28:54Z" level=info msg="rendering index \"registry.redhat.io/redhat/redhat-operator-index:v4.13\" as file-based catalog"
time="2024-10-11T11:29:49Z" level=info msg="wrote rendered file-based catalog to \"/tmp/knative.aIjkdvzz/tmp.IiMQGlnAOi\"\n"
11:30:27.698 INFO:    Generating catalog for OCP 4.14
time="2024-10-11T11:30:27Z" level=info msg="rendering index \"registry.redhat.io/redhat/redhat-operator-index:v4.14\" as file-based catalog"
time="2024-10-11T11:31:21Z" level=info msg="wrote rendered file-based catalog to \"/tmp/knative.aIjkdvzz/tmp.s4bv1h7HZi\"\n"
11:31:39.478 INFO:    Generating catalog for OCP 4.15
time="2024-10-11T11:31:39Z" level=info msg="rendering index \"registry.redhat.io/redhat/redhat-operator-index:v4.15\" as file-based catalog"
time="2024-10-11T11:32:31Z" level=info msg="wrote rendered file-based catalog to \"/tmp/knative.aIjkdvzz/tmp.vOKEBnWFRH\"\n"
11:32:48.077 INFO:    Generating catalog for OCP 4.16
time="2024-10-11T11:32:48Z" level=info msg="rendering index \"registry.redhat.io/redhat/redhat-operator-index:v4.16\" as file-based catalog"
time="2024-10-11T11:33:36Z" level=info msg="wrote rendered file-based catalog to \"/tmp/knative.aIjkdvzz/tmp.mKSK5ppXVI\"\n"
11:33:57.680 INFO:    Generating catalog for OCP 4.17
time="2024-10-11T11:33:57Z" level=info msg="rendering index \"registry.redhat.io/redhat/redhat-operator-index:v4.17\" as file-based catalog"
time="2024-10-11T11:34:12Z" level=info msg="wrote rendered file-based catalog to \"/tmp/knative.aIjkdvzz/tmp.TB6GDPKkiP\"\n"
```

pierDipi commented on olm-catalog/serverless-operator-index/v4.14/catalog/serverless-operator/catalog.json:
This one doesn't seem to be a valid JSON, there are missing commas

![image](https://github.com/user-attachments/assets/e42bf79c-e275-4ab2-b394-1b3eace48c3a)

pierDipi replied:
and I guess the same for others, but if that's normal, nevermind my noob questions

pierDipi replied:
In the same file I see old **CI** images

```
        {
            "name": "knative-openshift-ingress",
            "image": "registry.ci.openshift.org/knative/release-1.34.0:serverless-ingress"
        },
        {
            "name": "knative-openshift",
            "image": "registry.ci.openshift.org/knative/release-1.34.0:serverless-knative-operator"
        },
        {
            "name": "knative-operator",
            "image": "registry.ci.openshift.org/knative/release-1.34.0:serverless-openshift-knative-operator"
        },
```

pierDipi replied:
shouldn't there be prod images only in this file as that's part of the final release?

creydr replied:
I got the same, when I created them as described https://github.com/konflux-ci/olm-operator-konflux-sample/blob/main/docs/konflux-onboarding.md#create-the-fbc-in-the-git-repository

btw: those commands also offer a `--output yaml` parameter. So the catalog-templat might be easier to maintain later.

mgencur replied:
I think it doesn't matter that it's not a valid JSON. AFAIK, that's how it's generated by the `opm` tooling. And the catalog understands that.
I could try fixing it, but it works as it is.

mgencur replied:
I'm including 1.34 from CI because the catalog needs to have a complete graph. And right now 1.34 is not yet in registry.redhat.io nor in Konflux. 
This will be automatically fixed later. When 1.34 is pushed to the production catalog it will be pulled when calling "generate-catalog" and the script will skip this version and won't add CI images there.

mgencur replied:
One thing that is missing from this PR is replacing konflux reference to the bundle image with registry.redhat.io (after generating the catalog). And also deploying ImageContentSourcePolicy for this. But I will do this in a follow-up PR.

mgencur replied:
Trying to see if using yaml will be easier. I might push another commit later today.

pierDipi replied:
what about that comment ? https://github.com/openshift-knative/serverless-operator/pull/2926#discussion_r1793438797

pierDipi replied:
is it correct / what we expect ?

pierDipi replied:
I guess it's here the reply
> This will be automatically fixed later. When 1.34 is pushed to the production catalog it will be pulled when calling "generate-catalog" and the script will skip this version and won't add CI images there.

mgencur replied:
Yes. That's the answer.

mgencur replied:
Pushed the yaml variant ! I like it much more.

pierDipi commented on hack/lib/images.bash:
```diff
@@ -235,6 +245,21 @@ function latest_konflux_image_sha() {
   echo "${image_without_tag}@${digest}"
 }
 
+function latest_registry_ci_sha() {
+  input=${1:?"Provide image"}
+
+  image_without_tag=${input%:*} # Remove tag, if any
+
+  go_bin="$(go env GOPATH)/bin"
+  export GOPATH="$PATH:$go_bin"
+  digest=$(skopeo inspect --no-tags=true "docker://${input}" | jq -r '.Digest')
+  if [ "${digest}" = "" ]; then
+    exit 1
+  fi
+
+  echo "${image_without_tag}@${digest}"
+}
```
I think we could reuse `latest_konflux_image_sha` for this ?

mgencur replied:
I will check this. We don't want the "latest" tag though. We want a specific tag, e.g. release-1.34.0. The `latest_konflux_image_sha` automatically adds "latest"

mgencur replied:
Pushed.

creydr commented on hack/lib/images.bash:
```diff
@@ -235,6 +245,21 @@ function latest_konflux_image_sha() {
+  image_without_tag=${input%:*} # Remove tag, if any
+
+  go_bin="$(go env GOPATH)/bin"
+  export GOPATH="$PATH:$go_bin"
```
not sure, why this is needed?

mgencur replied:
I'm not sure. I copied it from the other functions in this file. Looks like it works without this as well. Maybe @pierDipi knows?

mgencur replied:
Removed the line in the new commit.

pierDipi replied:
not everyone has the go/bin folder in their path and `go install` installs binaries in go/bin but I'm ok making it a prerequisite

pierDipi commented on .github/workflows/validate.yaml:
```diff
@@ -62,6 +66,15 @@ jobs:
         working-directory: ./src/github.com/${{ github.repository }}
         run: make generated-files
 
+      # Generate catalog only if not triggered by a pull request. Credentials are not available in pre-submit phase.
+      - name: Regenerate file-based catalog
+        if: github.event_name == 'workflow_dispatch' || github.event_name == 'schedule' || ( github.event_name == 'push' && !contains(github.ref_name, 'dependabot/') )
+        working-directory: ./src/github.com/${{ github.repository }}
+        env:
+          REGISTRY_REDHAT_IO_USERNAME: ${{ secrets.REGISTRY_REDHAT_IO_USERNAME }}
+          REGISTRY_REDHAT_IO_PASSWORD: ${{ secrets.REGISTRY_REDHAT_IO_PASSWORD }}
+        run: make generate-catalog
```
I think we want to also run this on `bump` https://github.com/openshift-knative/serverless-operator/blob/da827643dd892789387b2c58b61e9ed97c2d907f/.github/workflows/bump.yaml#L91-L93

mgencur commented:
I built the catalog image locally and deployed to OpenShift, and it works!
![image](https://github.com/user-attachments/assets/059a70b8-ec68-4fc5-b976-f03276b0120c)

After building the image I just ran these two commands that install the CatalogSource.
```
export INDEX_IMAGE=quay.io/mgencur/serverless-index:release-1.35.0-4.17
./hack/dev.sh
```

pierDipi commented:
Awesome!

creydr commented:
Awesome :tada:

pierDipi commented:
/lgtm

mgencur commented:
> One thing that is missing from this PR is replacing konflux reference to the bundle image with registry.redhat.io (after generating the catalog). And also deploying ImageContentSourcePolicy for this. But I will do this in a follow-up PR.

I had to do a rebase because of some other changes so I added a couple of commits and fixed the above as well.

creydr reviewed (commented):
/lgtm

creydr commented:
/test 416-test-upgrade-aws-416
